File: spoiler-shield/ml/src/spoiler_shield/data/segment.py
"""Sentence splitting for training data.

Mirrors the extension's splitter (``extension/lib/segment.ts``): split at
sentence-ending punctuation, then merge pieces that end in a known
abbreviation or a single-letter initial. Keeping training and inference
segmentation alike matters, because the model only ever sees one sentence.
The browser uses ICU via ``Intl.Segmenter``; small differences on unusual
punctuation are expected and measured on the gold set.
"""

from __future__ import annotations

import re

_ABBREVIATION_WORDS = (
    "mr", "mrs", "ms", "dr", "prof", "st", "jr", "sr", "capt", "lt", "sgt", "gen", "col",
    "vs", "etc", "e.g", "i.e", "approx", "no", "vol", "ep", "eps", "pt", "ch", "fig", "mt",
)  # fmt: skip
ABBREVIATIONS = frozenset(f"{a}." for a in _ABBREVIATION_WORDS)

# A sentence ends at . ! ? optionally followed by one closing quote or bracket,
# then whitespace. Lookbehinds keep the punctuation and quote in the sentence.
_CLOSERS = "\"'\u2019\u201d)\\]"
_BOUNDARY = re.compile(rf"(?<=[.!?])\s+|(?<=[.!?][{_CLOSERS}])\s+")


def _ends_with_abbreviation(piece: str) -> bool:
    words = piece.split()
    last = words[-1].lower() if words else ""
    return last in ABBREVIATIONS or re.fullmatch(r"[a-z]\.", last) is not None


def split_sentences(text: str) -> list[str]:
    pieces = [p for p in _BOUNDARY.split(text) if p.strip()]
    out: list[str] = []
    carry = ""
    for piece in pieces:
        carry = f"{carry} {piece}" if carry else piece
        if not _ends_with_abbreviation(carry):
            out.append(carry.strip())
            carry = ""
    if carry.strip():
        out.append(carry.strip())
    return out

File: spoiler-shield/extension/lib/segment.ts
/**
 * Sentence splitting for page text.
 *
 * Uses the browser's built-in `Intl.Segmenter`, then repairs its main weakness:
 * ICU breaks after abbreviations such as "Dr." or "ep.", which would cut
 * "Dr. Mara Quinn was the informant." into two pieces and hide the context the
 * model needs. A piece ending in a known abbreviation or a single-letter
 * initial is merged with the piece that follows it.
 */

const ABBREVIATIONS = new Set(
  [
    'mr', 'mrs', 'ms', 'dr', 'prof', 'st', 'jr', 'sr', 'capt', 'lt', 'sgt', 'gen', 'col',
    'vs', 'etc', 'e.g', 'i.e', 'approx', 'no', 'vol', 'ep', 'eps', 'pt', 'ch', 'fig', 'mt',
  ].map((a) => `${a}.`),
);

function endsWithAbbreviation(piece: string): boolean {
  const lastWord = piece.trimEnd().split(/\s+/).pop()?.toLowerCase() ?? '';
  if (ABBREVIATIONS.has(lastWord)) return true;
  // Single-letter initial such as the "J." in "J. Doe".
  return /^[a-z]\.$/.test(lastWord);
}

export function splitSentences(text: string, locale = 'en'): string[] {
  const segmenter = new Intl.Segmenter(locale, { granularity: 'sentence' });
  const merged: string[] = [];
  let carry = '';
  for (const { segment } of segmenter.segment(text)) {
    carry += segment;
    if (!endsWithAbbreviation(carry)) {
      merged.push(carry);
      carry = '';
    }
  }
  if (carry) merged.push(carry);
  return merged.map((s) => s.trim()).filter((s) => s.length > 0);
}
